Repo signature rework

ydirson opened the pull request.

ydirson committed:
README: describe the workflow in more details

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
README: fix h2 titles capitalization inconsistencies

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
README: use github beta admonition syntax, not just an emoji

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

stormi commented on README.md:
```diff
@@ -60,14 +127,31 @@ This scripts excludes from the mirror:
+> [!NOTE]
+>
+> this is not a general-purpose mirroring tool.  It will
+> notably not mirror a number of deveopment packages, which you owuld
```
```suggestion
> notably not mirror a number of development packages, which you would
```

ydirson replied:
done

stormi reviewed (commented):
LGTM at this stage (I only checked the readme)

ydirson committed:
create-iso: avoid extracting branding info for netinstall images

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
mirror-repos: allow mirroring arbitrary URLs, improve its documentation

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
mirror-repos: filter out largest unused RPMs

Scales down current mirrors:
- (unreleased) 8.3 from 5.3GB to 2.7GB
- 8.2 from 5.5GB to 3.2GB

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Delegate key signing from create-iso to an exernal script

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
Rename create-install-iso to create-iso

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson commented:
Remaining issue is that the yum repo is set up in such a way that a single GPG key has to be used for both `repo-gpg-check` (checking the sig on repo metadata, which is the sig on which this PR focuses) and for `gpg-check` (checking the sig on individual RPMs).
My position is that the `gpg-check=1` setting is superfluous anyway: the sig on the repo metadata encompasses everything in the repo, down to the RPMs themselves. We could simply change it `gpg-check=0` without sacrificing anything security-wise with our install media (provided they checked the signature of the image itself before booting it, as it is the source of the public key used to verify the repo when doing netinstall).  Someone building a custom media ought to sign their repo metadata and gain the same level of integrity. If they don't (but, again, they really should), they can only get some level of integrity for netinstall through `gpg-check=1`, and this only if they add RPMs that are already signed by the key provided on the ISO -- so really this is a tiny niche, and they should sign their metadata instead.

stormi commented on README.md:
```diff
@@ -59,15 +95,55 @@ This scripts excludes from the mirror:
 - development RPMs
 - debugging-symbols RPMs
+- a few large RPMs only useful as build dependencies: ocaml, golang,
+  java
```
```suggestion
  java, xs-opam-repo, etc.
```

ydirson replied:
I considered `xs-opam-repo` to be part of `ocaml`

stormi replied:
ok

stormi reviewed (commented):
It looks good to me overall.

Minor regression: `create-iso.sh` with `--netinstall` adds a `.treeinfo` file (whose placeholders were not replaced)

ydirson commented:
> Minor regression: `create-iso.sh` with `--netinstall` adds a `.treeinfo` file (whose placeholders were not replaced)

Good catch, fixing that one

ydirson commented:
Only one issue remain (for signing with any other key than the XCP-ng one):
- boot with `no-gpgcheck` by default?
- keep it and tell the users to do that themselves (scripted in sample sign script)?
- dig further do find how to separate the repo and rpm sig checking, which may imply changes to host-installer?
- or?

stormi commented:
> Only one issue remain (for signing with any other key than the XCP-ng one):
> 
>     * boot with `no-gpgcheck` by default?
>     * keep it and tell the users to do that themselves (scripted in sample sign script)?
>     * dig further do find how to separate the repo and rpm sig checking, which may imply changes to host-installer?
>     * or?

I like the idea to let the sample script disable gpgcheck. But then, will signing the repodata serve any purpose? 

Best option in the long term is probably option 3, but we can let it wait for when either users or ourselves express a need for it.

ydirson commented:
> I like the idea to let the sample script disable gpgcheck. But then, will signing the repodata serve any purpose?

You indeed close the door on the idea here :smile: 
I was thinking 8.3 where we have separate controls for `gpgcheck` and `repo_gpgcheck`, but the original xcpng patches in 8.2 have a single control for both, so that cannot work on our LTS version.

That would require the sample script to behave differently depending on the version, that's IMHO not the place for it.

> Best option in the long term is probably option 3, but we can let it wait for when either users or ourselves express a need for it.

Until we can dedicate some time for 3., that seems to leave only 1., with a new config file in `create-iso` to let 8.3 pass `no-gpgcheck` and not modify the 8.2 behavior (after all, 8.2 never supported re-signing the repo without resigning the RPMs, it seems).

ydirson committed:
create-iso: factorize code modifying bootloader configs

Copies had already diverged.

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson committed:
create-iso: add boot args from new installer-bootargs.lst config files

Signed-off-by: Yann Dirson <yann.dirson@vates.fr>

ydirson commented:
Added the mechanism for 8.3 to use `no-gpgcheck`

stormi commented:
LGTM. I'll generate a signed ISO one last time before I approve the PR.

@gduperrey as planned by Yann and myself, this pull request changes the way we call the generation scripts somehow. Not as much as we initially planned, but maybe enough to require adapting the CI jobs that generate daily (unsigned) ISOs. Please test them and adapt them if needed, with Yann's help.

Once you have Before/After ISOs for both 8.2 and 8.3, I'll compare them and look for unexpected changes.

Then we'll merge this PR.

gduperrey commented:
@stormi I have a PR pending.
As Jenkins is using a local copy, I will copy/manage it when the PRs will be merged.

stormi commented:
Unless I've overlooked it, I think the README is lacking an actual example of invocation of `create-iso.sh` using a signing script.

stormi reviewed (changes requested):
I generated a 8.3:updates ISO, signed, but the bootloader entries all contain `no-gpgcheck`.

ydirson commented:
Changed default behavior not to add `no-gpgcheck` and let users add it in their own layer when needed.
